Security analysis of the relay protocol

staheri14 opened the pull request:
This PR addresses the following issues 
https://github.com/vacp2p/specs/issues/239
https://github.com/vacp2p/specs/issues/244

The summary of the edits:
- A list of the security requirements (this is an initial list, more items can be added as we move forward)
- The description of the adversarial model (for now a passive adversary is considered, future analysis can cover an active adversary as well)
- The initial security assessment of the current version of the relay protocol (will certainly add more detailed analysis in future PRs)

staheri14 committed:
WIP: adds the security analysis file with some initial  incomplete writeup

staheri14 committed:
WIP: adds the security analysis file with some initial  incomplete writeup

staheri14 committed:
fixes some formatting issue

staheri14 committed:
fixes some formatting issue

staheri14 committed:
minor

staheri14 committed:
minor

staheri14 committed:
changes a terminology

staheri14 committed:
changes a terminology

staheri14 committed:
adds the outline

staheri14 committed:
adds the outline

staheri14 committed:
Anlysis -> Analysis

staheri14 committed:
Anlysis -> Analysis

staheri14 committed:
removes old changelogs

staheri14 committed:
removes old changelogs

staheri14 committed:
edits the publisher anonymity

staheri14 committed:
edits the publisher anonymity

staheri14 committed:
Minor

staheri14 committed:
Minor

oskarth commented on specs/waku/v2/waku-relay-security-analysis.md:
```diff
@@ -0,0 +1,108 @@
+
+# Abstract
+
+In this spec, we analyze the security of the  `relay` protocol concerning data confidentiality, integrity, authenticity, and anonymity. This is to enable users of this protocol to make informed decision about all the secuirty properties that can or can not achieve by deploying `relay` protocol. 
```
I think this should be part of the existing relay spec, we want to be clear about guarantees in that spec, not just analyze it from a far with a different spec (which also can't really be implemented, more of an analysis/guarantees)

staheri14 replied:
> I think this should be part of the existing relay spec, we want to be clear about guarantees in that spec, not just analyze it from a far with a different spec (which also can't really be implemented, more of an analysis/guarantees)

Sure, addressed!

oskarth commented on specs/waku/v2/waku-relay-security-analysis.md:
```diff
@@ -0,0 +1,108 @@
+Personally identifiable information (PII): PII indicates any piece of data that can be used to uniquely identify a Peer. For example, the signature verification key, and the hash of one's IP address are unique for each peer and hence count as PII.
+
+## Anonymity
+### Definition
```
Threat model and definition seems like good sections in the main relay spec

staheri14 replied:
> Threat model and definition seems like good sections in the main relay spec

I embedded the security analysis into the relay specs

oskarth commented on specs/waku/v2/waku-relay-security-analysis.md:
```diff
@@ -0,0 +1,108 @@
+    - [Analysis](#analysis-2)
+  - [Confidentiality](#confidentiality)
+    - [Definition](#definition-3)
+    - [Analysis](#analysis-3)
```
I'd probably make this a bit tighter.

Here are some examples that might give inspiration in terms of structure:

https://code.briarproject.org/briar/briar-spec/-/blob/master/protocols/BHP.md and https://code.briarproject.org/briar/briar-spec/-/blob/master/protocols/BTP.md

staheri14 replied:
Thanks! changed the structure

oskarth reviewed (commented):
I know it is a draft, just thought I'd leave some initial comments :)

staheri14 committed:
incorporates the security section into the relay specs

staheri14 committed:
incorporates the security section into the relay specs

staheri14 committed:
updates authors

staheri14 committed:
updates authors

staheri14 committed:
deletes the security analysis file

staheri14 committed:
deletes the security analysis file

staheri14 committed:
adds the security requirements

staheri14 committed:
adds the security requirements

staheri14 committed:
adds the adversarial model

staheri14 committed:
adds the adversarial model

staheri14 committed:
adds security analysis and future work

staheri14 committed:
adds security analysis and future work

staheri14 committed:
adds a todo

staheri14 committed:
adds a todo

staheri14 committed:
removes an empty item

staheri14 committed:
removes an empty item

staheri14 commented:
> I know it is a draft, just thought I'd leave some initial comments :)

Thanks a lot for your great comments!

staheri14 committed:
Update waku-relay.md

staheri14 committed:
Update waku-relay.md

staheri14 committed:
adds the terminologies subsection

staheri14 committed:
adds the terminologies subsection

staheri14 committed:
Merge remote-tracking branch 'origin/security-analysis/relay' into security-analysis/relay

staheri14 committed:
adds some revisions

staheri14 committed:
adds some revisions

staheri14 committed:
adds some todos

staheri14 committed:
adds some todos

staheri14 committed:
adds more todos

staheri14 committed:
adds more todos

staheri14 committed:
mentions the traffic analysis attack

staheri14 committed:
mentions the traffic analysis attack

staheri14 committed:
minor edits on the adversarial model

staheri14 committed:
minor edits on the adversarial model

staheri14 committed:
adds some clarification about the adversarial model & some final edits

staheri14 committed:
adds some clarification about the adversarial model & some final edits

staheri14 committed:
changes topicID to topic ID

staheri14 committed:
changes topicID to topic ID

staheri14 committed:
Update waku-relay.md

staheri14 committed:
Update waku-relay.md

staheri14 committed:
adds minor clarification

staheri14 committed:
adds minor clarification

jm-clius commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +30,34 @@ authors: Oskar Thorén <oskar@status.im>
+-  **Authenticity**: No adversary can forge data on behalf of a targeted peer and make it accepted by other peers as if the origin is the target.
+- **Spam resistant**: No adversary can flood the system with spam messages (i.e., publishing a large number of messages in a short amount of time).
+  
+<!-- TODO: more requirements can be added, but that needs further and deeper investigation-->
```
For interest's sake, is this where the resilience of the underlying GossipSub protocol to attacks from malicious nodes come in? [This research article](https://www.researchgate.net/publication/342734066_GossipSub_Attack-Resilient_Message_Propagation_in_the_Filecoin_and_ETH20_Networks) lists a number of attacks and mitigation strategies, e.g. Sybil nodes affecting message propagation, mesh creation, etc. As far as I can tell many of the suggested mitigation strategies are implemented in nim-libp2p, though there may still be some vulnerabilities.

staheri14 replied:
> This research article lists a number of attacks and mitigation strategies, e.g. Sybil nodes affecting message propagation, mesh creation, etc. As far as I can tell many of the suggested mitigation strategies are implemented in nim-libp2p, though there may still be some vulnerabilities.

Thanks! will check it out

> is this where the resilience of the underlying GossipSub protocol to attacks from malicious nodes come in?

Yes, that is right

staheri14 replied:
> As far as I can tell many of the suggested mitigation strategies are implemented in nim-libp2p

Could not find the 6 scoring factors of the scoring function in nim-libp2p as described in the article, do you have any idea about their corresponding implementations in the codebase?

jm-clius reviewed (approved):
LGTM

staheri14 commented:
>  For interest's sake, is this where the resilience of the underlying GossipSub protocol to attacks from malicious nodes come in? [This research article](https://www.researchgate.net/publication/342734066_GossipSub_Attack-Resilient_Message_Propagation_in_the_Filecoin_and_ETH20_Networks) lists a number of attacks and mitigation strategies, e.g. Sybil nodes affecting message propagation, mesh creation, etc. As far as I can tell many of the suggested mitigation strategies are implemented in nim-libp2p, though there may still be some vulnerabilities.

Thanks @jm-clius!
I read the paper and would like to share a couple of points about the attacks and strategies discussed in the article:

- The peer scoring method can be used for spam protection and I already included it as a related work for the comparison (under a todo item in the specs). However, the scoring function as defined in the article, and as far as I understood, does not explicitly cover the spamming scenario (but waku rln-relay is going to address the spamming issue as well). 


- The security analysis covered in the current PR addresses the static adversarial model whereas the attacks discussed in the article mainly fit the active adversarial model (e.g., nodes that may lie about the messages they get). As such, those attacks are not included in the current PR and will be analyzed in future issues.

- The attacks discussed in the GossipSub article are all due to Sybil nodes, and as far as I understood (certainly need more investigation), the paper does not provide any strategy to prevent the creation of Sybil nodes but instead, it proposes the scoring function to alleviate the effect of those attacks (detect misbehaviors). In contrast, in the rln relay (which is the solution for protection against the spammers in the waku rln-relay), id creation comes with a financial cost, as such, it is expected that we will experience less Sybil nodes (intuitively, but has to be compared practically). And if we prove so, then waku rln-relay is naturally protected against the attacks mentioned in the Gossipsub article (this analysis also deserves its separate issue)


- Finally, the nim-libp2p codebase needs to be audited to make sure that all the strategies listed in the paper are translated to the nim version as well (this should be also addressed in a separate issue).

oskarth commented on specs/waku/v2/waku-relay.md:
Hm, something weird happened here. We are already on beta2.

Would be good to add a one-liner to `### Next`, with something like "added initial threat model, security analysis ~". Then for beta3 we just change `Next` to beta3 and it is easier to see how spec has evolved over time.

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +30,34 @@ authors: Oskar Thorén <oskar@status.im>
+
+- **Confidentiality**: The adversary should not be able to learn the data carried by the `relay` protocol.
+-  **Integrity**: This feature indicates that the data transferred by the `relay` protocol can not be tampered with by an adversarial entity without being detected.
+-  **Authenticity**: No adversary can forge data on behalf of a targeted peer and make it accepted by other peers as if the origin is the target.
```
Is this accurate, consider we do message signing?

staheri14 replied:
What does concern you about accuracy? If we sign messages then we preserve authenticity (should be careful about replay attack though)

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +30,34 @@ authors: Oskar Thorén <oskar@status.im>
 
 **Protocol identifier***: `/vac/waku/relay/2.0.0-beta2`
 
+## Security Requirements
```
I wonder if this is "security requirements" or if something like "design goals" would make sense. How do you think about this distinction? Aspirational/design vs requirements <for something>

staheri14 replied:
I agree, "security requirements" might be misleading as the reader may think these are addressed in the protocol, but I feel the same about the "design goals", it sounds like the current design covers all the listed security features (which is not the case). The word "Design" is also general and includes other aspects like performance as well. What about "Security Aspects"?

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +30,34 @@ authors: Oskar Thorén <oskar@status.im>
+  
+<!-- TODO: more requirements can be added, but that needs further and deeper investigation-->
+
+### Terminologies
```
Would it make sense to also distinguish Operator from Peer? E.g. an operator/user can have one or many peers, and on or many node IDs. A pseudonymous platform would make it hard to tie an individual id to a specific human (e.g.), but something like phone number/email makes this harder. So even if you can see a node id in a public chat (e.g.) it protects IP, as well as human operator. See phrasing in https://our.status.im/our-principles/

> IV. Privacy
Privacy is the power to selectively reveal oneself to the world. For us, it's essential to protect privacy in both communications and transactions, as well as being a pseudo-anonymous platform. Additionally, we strive to provide the right of total anonymity.

staheri14 replied:
Oh, I see your point! Will change it to an individual instead of Peer.

Regarding the pseudonymity, would like to share some thoughts:
To the best of my knowledge, Pseudonymity does not imply and guarantee anonymity (as you mentioned). Indeed, in may fields it is practically shown that masking individual's identity through pseudonyms e.g., PeerID in our case, does not prevent re-identifying that individual (i.e., does not preserve unlink-ability) and as you said, pseudonymity only makes the linkability "hard" (while "hard" sometimes is practically shown to be "feasible" and "easy"). Will get back to you with some sample resources in this regard.
Just to give an intuition, PeerIDs are like users footprint, hence all the activity associated with that footprint can be used to profile one single individual and finally such profile combined with some external information would lead to the re-identification risk. 
Having multiple Peer IDs is like partitioning the users activity into disjoint sets, but the same profiling attack may be still valid in this case (should be analyzed and researched).
In summary, to achieve a higher level of anonymity, it is better to hide the PII as much as possible.

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +30,34 @@ authors: Oskar Thorén <oskar@status.im>
+
+## Adversarial Model
+-  Any peer talking the `relay` protocol is considered as an adversary. This includes publishers, subscribers, and all the peers' direct connections. Furthermore, we consider the adversary as a passive entity that attempts to collect information from other peers to conduct an attack but it does so without violating protocol definitions and instructions. For example, under the passive adversarial  ...
+- The following are not considered as part of the adversarial model: 1- An adversary with a global view of all the peers and their connections 2- An adversary that can eavesdrop on communication links between arbitrary pair of peers (unless the adversary is one end of the communication). In specific, the communication channels are assumed to be secure.
```
I'd make this a list on each newline

staheri14 replied:
sure

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -83,13 +114,30 @@ The `subscribe` field MUST contain a boolean, where 1 means subscribe and 0 mean
+## Security Analysis 
+<!-- TODO: realized that the prime security objective of the `relay` protocol is to provide peers anonymity as such this feature is prioritized over other features e.g., anonymity is preferred over authenticity and integrity. It might be good to motivate anonymity and its impact on the relay protocol or other protocols invoking relay protocol.-->
+
+- **Message Publisher Anonymity**: To preserve message publisher anonymity, one should remove any PII from the published message. As such, `relay` protocol follows the `StrictNoSign` policy as described in [libp2p PubSub specs](https://github.com/libp2p/specs/tree/master/pubsub#message-signing). As the result of the `StrictNoSign` policy, `Message`s should be built without the  `from`, `signature ...
```
Great

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -83,13 +114,30 @@ The `subscribe` field MUST contain a boolean, where 1 means subscribe and 0 mean
+
+## Future work
+
+- **Confidentiality**: The current version of the `relay` protocol does not address confidentiality, however, it is achievable by performing encryption on top of the `data` field of `Message`s. More details on enabling encryption mode are provided in [Libp2p PubSub specs](https://github.com/libp2p/specs/tree/master/pubsub#the-topic-descriptor) as part of the `TopicDescriptor` messages.
```
Is this in scope of the relay protocol, or is this more something that we want to defer to Waku Message spec? So more of an informational note perhaps? Ditto below

staheri14 replied:
I agree that confidentiality of the `data` can be addressed at different protocol layers and in this case the `data` field which is a waku message could be encrypted before being passed to the `relay` protocol, however, since the `relay` protocol carries the `data` field, the decision of whether the `data`  should be encrypted/sealed or not could be (and is good to be) part of this protocol as well  (moreover, the encryption option is already offered by libp2p using `TopicDescriptor`s) 
Of course, if no encryption takes place at the `relay` level, one can always encrypt the waku message before handing to the `relay` protocol.
Please let me know your thoughts.

oskarth reviewed (commented):
Generally looks great, especially for a timeboxed attack, some comments and stuff in line

staheri14 committed:
rephrases the PII

staheri14 committed:
itemizes the adversarial model

staheri14 committed:
adds the next changelog

staheri14 committed:
Update waku-relay.md

staheri14 committed:
restructures, removes unsupported features, adds further clarifications

staheri14 committed:
Update waku-relay.md

staheri14 committed:
fixes some wordings

staheri14 committed:
comments out footnotes

staheri14 committed:
substitutes relay with wakurelay

staheri14 committed:
minor

staheri14 committed:
fixes a notational issue

staheri14 committed:
Update waku-relay.md

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -7,6 +7,7 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
 
 # Table of Contents
 
+- [Table of Contents](#table-of-contents)
```
Extra

staheri14 replied:
I think you reviewed an old version of the branch (an outdated waku-relay.md with this commit https://github.com/vacp2p/specs/pull/268/commits/25d22970a0f657464a4db3ac90a722eb47b4c298), the branch had further commits and updates after that

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
 ## Security Requirements
 
-<!-- In this part, we analyze the security of the  `relay` protocol concerning data confidentiality, integrity, authenticity, and anonymity. This is to enable users of this protocol to make an informed decision about all the security properties that can or can not achieve by deploying a `relay` protocol.-->
+The `relay` protocol supports the following security properties under a static [Adversarial Model](#adversarial-model). Note that data confidentiality[^1], integrity[^2] and authenticity[^3] are currently considered out of the scope of `relay` protocol and must be handled by the upperlevel protocols. 
```
upper level protocols, such as <Waku Message>

staheri14 replied:
Message encryption and message signatures should be done at the time of message generation (i.e., in some upper-level protocol). However,  among the stack of protocols we have (in waku v2) , I could not find one in which message generation takes place (please let me know if there is such protocol) e.g., the filter and store protocols only read and operate on the messages received by the waku relay but no peer can publish waku message as part of these protocols.

I think that upper level protocols stand for those that use waku v2 as their messaging protocol.

oskarth replied:
sorry that was meant as a typo fix: "upper level protocols, such as <Waku Message>, etc" (not upperlevel, and give pointer to waku message)

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
 ## Security Requirements
 
-<!-- In this part, we analyze the security of the  `relay` protocol concerning data confidentiality, integrity, authenticity, and anonymity. This is to enable users of this protocol to make an informed decision about all the security properties that can or can not achieve by deploying a `relay` protocol.-->
+The `relay` protocol supports the following security properties under a static [Adversarial Model](#adversarial-model). Note that data confidentiality[^1], integrity[^2] and authenticity[^3] are currently considered out of the scope of `relay` protocol and must be handled by the upperlevel protocols. 
```
supports vs "is designed for"?

staheri14 replied:
Changed it to "is designed"

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
-## Adversarial Model
--  Any peer talking the `relay` protocol is considered as an adversary. This includes publishers, subscribers, and all the peers' direct connections. Furthermore, we consider the adversary as a passive entity that attempts to collect information from other peers to conduct an attack but it does so without violating protocol definitions and instructions. For example, under the passive adversarial  ...
+## Adversarial Model 
+-  Any entity talking the `relay` protocol is considered as an adversary. This includes publishers, subscribers, and all the peers' direct connections. Furthermore, we consider the adversary as a passive entity that attempts to collect information from others to conduct an attack but it does so without violating protocol definitions and instructions. For example, under the passive adversarial mod ...
```
A node running the relay protocol perhaps? Instead of "talking"

staheri14 replied:
addressed

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
-- **Spam resistant**: No adversary can flood the system with spam messages (i.e., publishing a large number of messages in a short amount of time).
+- **Publisher-Message Unlinkability**: This property indicates that no adversarial entity can link a published `Message` to its publsiher. This feature also implies the unlinkability of the publisher to its published topic ID as the `Message` embodies the topic IDs.
+
+- **Subscriber-Topic Unlinkability**: This feature stands for the inability of any adversarial entity from linking a subscriber to its subscribed topic IDs.
```
Is this accurate? If I run a node and you subscribe to a topic, doesn't this mean I see what topics you are subscribed to?

staheri14 replied:
Following our last meeting, since there is only one default topic id (pubsub topic) in waku v2, such un-linkability is guaranteed by default (it provides k-anonymity where k is the size of the system). This is detailed in the security analysis as well.

oskarth replied:
Note that one topic applies to _Waku v2_ spec, not to _Waku Relay_ spec.

oskarth replied:
>  only one default topic id (pubsub topic) in waku v2

This is how Waku v2 is using it, but it isn't what the relay protocol provides. The relay protocol enables the use of many topics, but our recommended usage in Waku v2 spec is to use a single one.

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
--  **Integrity**: This feature indicates that the data transferred by the `relay` protocol can not be tampered with by an adversarial entity without being detected.
--  **Authenticity**: No adversary can forge data on behalf of a targeted peer and make it accepted by other peers as if the origin is the target.
-- **Spam resistant**: No adversary can flood the system with spam messages (i.e., publishing a large number of messages in a short amount of time).
+- **Publisher-Message Unlinkability**: This property indicates that no adversarial entity can link a published `Message` to its publsiher. This feature also implies the unlinkability of the publisher to its published topic ID as the `Message` embodies the topic IDs.
```
publisher typo

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
+
 ### Terminologies
-The term Personally identifiable information (PII) refers to any piece of data that can be used to uniquely identify a user. For example, the signature verification key, and the hash of one's IP address are unique for each user and hence count as PII.
+The term Personally identifiable information (PII) refers to any piece of data that can be used to uniquely identify a user. For example, the signature verification key, and the hash of one's static IP address are unique for each user and hence count as PII.
```
what's the significance of static IP here? wouldn't dynamic IPs ocunt as potential PII?

staheri14 replied:
In the case of static IP address, there is a one to one mapping between the user and her IP, whereas in the dynamic case, an IP is attributable to multiple users (many to one mapping) therefore an IP cannot uniquely be linked to one single user (it preserves some level of k-anonymity where k is the number of users sharing the same IP). While a dynamic IP address alone does not count PII,  it may result in the linkability risk if combined with other available information.

oskarth replied:
fair!

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -119,21 +122,21 @@ The `topicid` field MUST contain the topic.
-
-- **Message Publisher Anonymity**: To preserve message publisher anonymity, one should remove any PII from the published message. As such, `relay` protocol follows the `StrictNoSign` policy as described in [libp2p PubSub specs](https://github.com/libp2p/specs/tree/master/pubsub#message-signing). As the result of the `StrictNoSign` policy, `Message`s should be built without the  `from`, `signature ...
-Note that removing identifiable information from messages cannot lead to perfect anonymity. The direct connections of a publisher might be able to figure out which `Message`s belong to that publisher by analyzing its traffic. The possibility of such inference may get higher when the `data` field is also not encrypted. <!-- TODO: more investigation on traffic analysis attacks and their success pro ...
+<!-- TODO: realized that the prime security objective of the `relay` protocol is to provide peers unlinkability as such this feature is prioritized over other features e.g., unlinkability is preferred over authenticity and integrity. It might be good to motivate unlinkability and its impact on the relay protocol or other protocols invoking relay protocol.-->
```
Agree, I think this would belong in a Motivation section at top

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -119,21 +122,21 @@ The `topicid` field MUST contain the topic.
 
-## Future work
+- **Publisher-Message Unlinkability**:  To address publisher-message unlinkability, one should remove any PII from the published message. As such, `relay` protocol follows the `StrictNoSign` policy as described in [libp2p PubSub specs](https://github.com/libp2p/specs/tree/master/pubsub#message-signing). As the result of the `StrictNoSign` policy, `Message`s should be built without the  `from`, `s ...
+Note that removing identifiable information from messages cannot lead to perfect unlinkability. The direct connections of a publisher might be able to figure out which `Message`s belong to that publisher by analyzing its traffic. The possibility of such inference may get higher when the `data` field is also not encrypted by the upper-lelvel protocols. <!-- TODO: more investigation on traffic anal ...
```
```suggestion

Note that removing identifiable information from messages cannot lead to perfect unlinkability. The direct connections of a publisher might be able to figure out which `Message`s belong to that publisher by analyzing its traffic. The possibility of such inference may get higher when the `data` field is also not encrypted by the upper-lelvel protocols. <!-- TODO: more investigation on traffic analysis attacks and their success probability-->
```

oskarth replied:
Link to Waku Message spec perhaps

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -119,21 +122,21 @@ The `topicid` field MUST contain the topic.
+Note that removing identifiable information from messages cannot lead to perfect unlinkability. The direct connections of a publisher might be able to figure out which `Message`s belong to that publisher by analyzing its traffic. The possibility of such inference may get higher when the `data` field is also not encrypted by the upper-lelvel protocols. <!-- TODO: more investigation on traffic anal ...
 
-- **Confidentiality**: The current version of the `relay` protocol does not address confidentiality, however, it is achievable by performing encryption on top of the `data` field of `Message`s. More details on enabling encryption mode are provided in [Libp2p PubSub specs](https://github.com/libp2p/specs/tree/master/pubsub#the-topic-descriptor) as part of the `TopicDescriptor` messages.
+- **Subscriber-Topic Unlinkability:** The `relay` protocol operates over only one topic id, as such, the subscriber-topic unlinkability is an immidiate property of this protocol. More specifically, subscribers are not reidentifiable from their subscribed  topic ids as the entire network is subscribed to the same topic id. This level of unlinkability / anonymity is known as [k-anonymity](https://w ...
```
I'm not sure this is accurate, the relay protocol itself doesn't stipulate how to use topics. This is a choice we make at the Waku v2 spec level to use one topic.

At the relay level, it is however a recommendation to "share topics" in order to leak minimum information (worded in a better way)

oskarth replied:
Maybe can be phrased as RECOMMEND to use limited topics, see Waku v2 spec for details on how to use relay protocol as a whole (or something like this?)

oskarth commented on specs/waku/v2/waku-relay.md:
This is partly supported by use of GossipSub 1.1, so I'd say it is a property that exists now but can be enhanced later. I.e. future work would instead be **Improved spam resistance** or **Economic spam resistance** or better protection again sybil agains, or similar

oskarth commented on specs/waku/v2/waku-relay.md:
simultaneously typo

oskarth replied:
interesting, I wonder how feasible/desirable this is

staheri14 replied:
> interesting, I wonder how feasible/desirable this is
 
It is an initial proposal, certainly needs further investigation to check the feasibility

oskarth commented on specs/waku/v2/waku-relay.md:
I'd mention peer scoring briefly. PoW would fit better in a section "Whisper/Waku v1 vs Waku v2"

staheri14 replied:
sure, done

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
+The `relay` protocol supports the following security properties under a static [Adversarial Model](#adversarial-model). Note that data confidentiality[^1], integrity[^2] and authenticity[^3] are currently considered out of the scope of `relay` protocol and must be handled by the upperlevel protocols. 
 
-- **Message Publisher Anonymity**: This property indicates that no adversarial entity can link a published `Message` to its origin i.e., the peer. Note that this feature also implies the unlinkability of the publisher to its published topic ID, this is because  `Message` contains the topic IDs as well.
+[^1]: Confidentiality indicates that an adversary should not be able to learn the data carried by the `relay` protocol.
```
I'd put all such references at the very end of the document. Same goes for external links, see eg https://specs.vac.dev/specs/waku/v2/waku-v2.html#references

staheri14 replied:
Thanks! sure (in the updated branch these are removed).

oskarth commented on specs/waku/v2/waku-relay.md:
```diff
@@ -33,26 +34,28 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
   
 <!-- TODO: more requirements can be added, but that needs further and deeper investigation-->
 
+
```
I'd call the Terminology or Definitions

staheri14 replied:
addressed

oskarth reviewed (commented):
In general looks good, structure seems reasonable to me. Can also iterate further.

I think it'd be useful to also have some links/analysis on Waku v2 https://specs.vac.dev/specs/waku/v2/waku-v2.html and Waku Message spec level. The reason for this is that people are going to compare Waku with Whisper, and in the original Whisper spec all of these properties were provided in a single spec. For us, relay protocol is just a part of package, since we have decoupled things a bit.

decanus committed:
Merge branch 'master' into security-analysis/relay

staheri14 committed:
WIP: addresses some of the comments

staheri14 committed:
Merge remote-tracking branch 'origin/security-analysis/relay' into security-analysis/relay

decanus commented on specs/waku/v2/waku-relay.md:
```diff
@@ -22,11 +31,39 @@ authors: Oskar Thorén <oskar@status.im>
 
 **Protocol identifier***: `/vac/waku/relay/2.0.0-beta2`
 
+## Security Requirements
```
Should this really be part of the abstract? right now the way the document is set up it is.

staheri14 replied:
Thanks for pointing this out! Apparently, everything in the relay specs is under the abstract section, will fix it.

staheri14 commented:
> I think it'd be useful to also have some links/analysis on Waku v2 https://specs.vac.dev/specs/waku/v2/waku-v2.html and Waku Message spec level. 

@oskarth  Sure, Will update those specs as well.

staheri14 committed:
WIP: clarifies upper level protocols


WIP: fixes headings issue


WIP: adds scoring function 


minor edits


updates the subscriber-topic unlinkability

staheri14 committed:
Merge branch 'master' into security-analysis/relay

decanus commented on specs/waku/v2/waku-relay.md:
```diff
@@ -16,6 +16,7 @@ authors: Oskar Thorén <oskar@status.im>, Sanaz Taheri <sanaz@status.im>
   - [RPC](#rpc)
   - [Message](#message)
   - [SubOpts](#subopts)
+  * [Signature Policy](#signature-policy)
```
```suggestion
  - [Signature Policy](#signature-policy)
```

staheri14 committed:
Update specs/waku/v2/waku-relay.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>

oskarth reviewed (approved):
This PR is getting quite big, and it definitely better than what we have so I'll go ahead and approve it

staheri14 committed:
Merge branch 'master' into security-analysis/relay
